### ISSUE TITLE
Destructive evaluation breaks for loops

### ISSUE THREAD

**Issue by vihanb** [cheddar-lang/Cheddar#3]
**Version**: v0.0.317-alpha

Simply: when evaluating an expression, the interpreter uses destructive commands to step. This means the expression is un-reevaluatable after it has been evaluated once. This is problematic to for loops in the fact they need to reevaluate the condition and the afterthought. An example is:

```js
for (var i := 0; i < 5; i=i+1) {
    print i
}
```

The output is:

```
0
ReferenceError: <blah blah blah, scary error etc.>
```

---

This is rather problematic as either A) the entire call stack of the expression would need to be cloned (which would be extremely slow). The other option would be to make the evaluation non-destructive which **means a complete rewrite.** Which I'm not looking forward too

**Comment by vihanb** [comment]
One fix is to attempt to pass a delegate array, where `Symbol.species` is faked as an `Array`. That way, this delegate array could extend the `Array` class, and abstract, and override the destructive methods with non-destructive ones.

This is still not the most desirable though as the delegate would be much slowing than natively using `.shift()` `.pop()`. As `unshift()` would require reallocating the entire block of memory allocated to the array

**Comment by vihanb** [comment]
(looks like in the refactoring of the git workflow, I lost the delegate array code :|) I still have the compiled version though

**Comment by vihanb** [comment]
```js
// Delegate Array
//  same as a normal array
//  but is a delegate in
//  the means it's non-
//  destructive

class DelegateArray extends Array {

    constructor(src) {
        super(...src); // So the delegate is views as an array

        if (src instanceof DelegateArray) {
            src._lp = 0;
            src._up = 0;
        } else {

            // Spread the source properties into here
            // not creating a closure as those are slow
            this._lp = 0; // Lower stack pointer
            this._up = src.length >>> 0; // Upper stack pointer
        }
    }

    // Fake behavior of destructive commands
    // Simply reset stack pointers to restore
    // the previous stack data

    // The movement of the stack is relative to
    // whether or not a head exists. Due to this
    // to verify the head:
    //   this._lp <= this._lr

    // Fake the destructive functions


    unshift(val) {
        // Lowers the bottom pointer, setting the
        // reference _and_ the pointer to a new
        // location. This way a new pointer will
        // be designated to the value but the
        // entire array does not have to be
        // reallocated
        this[--this._lp] = val;
    }

    shift(val) {
        // Return ths item at the lower reference
        // if it doesn't exist, this will return
        // the lower pointer location
        return this[this._lp++];
    }
};
```

**Comment by RikerW** [comment]
CAN YOU PLEASE MAKE ONE BIG COMMENT

5 emails while trying to play tf2 is not fun. >_>

**Comment by vihanb** [comment]
Fixed in v0.2.0: cf71c0660c1bd731a153e5914d94a293a41e5634
